llpc/lower/llpcSpirvLowerRayTracingIntrinsics.cpp

```diff
@@ -183,4 +211,108 @@ void SpirvLowerRayTracingIntrinsics::createConvertF32toF16(Function *func, unsig
+// Create AmdExtGetBaseAddrFromResource
+//
+// @param func : Function to create
+void SpirvLowerRayTracingIntrinsics::createGetBaseAddrFromResource(Function *func) {
+  // uint64_t AmdExtGetBaseAddrFromResource(resource)
+  // {
+  //   return resource.baseAddr;
+  // }
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Value *zero = m_builder->getInt32(0);
+
+  Value *resourcePtr = m_builder->CreateLoad(argIt->getType()->getPointerElementType(), argIt);
+  Value *ptr = m_builder->CreateGEP(resourcePtr->getType()->getPointerElementType(), resourcePtr, {zero});
+
+  m_builder->CreateRet(m_builder->CreatePtrToInt(ptr, m_builder->getInt64Ty()));
+}
+
+// =====================================================================================================================
+// Create AmdExtAtomicFMinAtAddrx2, AmdExtAtomicFMinAtAddr, AmdExtAtomicFMaxAtAddrx2, AmdExtAtomicFMaxAtAddr
+//
+// @param func : Function to create
+// @param isMin : Whether is Min operation, otherwise Max operation
+// @param is64Ty : Whether is 64-bit type operation, otherwise 32-bit type operation
+void SpirvLowerRayTracingIntrinsics::createAtomicFMinMaxAtAddr(Function *func, bool isMin, bool is64Ty) {
+  // float(2) AmdExtAtomicFMinAtAddr(x2)(uint64_t gpuVa, uint offset, float(2) value)
+
+  assert(func->getBasicBlockList().size() == 1);
+  (*func->begin()).eraseFromParent();
+
+  BasicBlock *entryBlock = BasicBlock::Create(m_builder->getContext(), "", func);
+  m_builder->SetInsertPoint(entryBlock);
+  auto argIt = func->arg_begin();
+
+  Type *gpuVaTy = m_builder->getInt64Ty();
+  Type *valueTy = is64Ty ? cast<Type>(FixedVectorType::get(m_builder->getFloatTy(), 2)) : m_builder->getFloatTy();
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), gpuVaTy));
+  Value *gpuVa = m_builder->CreateLoad(gpuVaTy, argIt++);
+  Value *offset = m_builder->CreateLoad(m_builder->getInt32Ty(), argIt++);
+  // TODO: Remove this when LLPC will switch fully to opaque pointers.
+  assert(IS_OPAQUE_OR_POINTEE_TYPE_MATCHES(argIt->getType(), valueTy));
+  Value *value = m_builder->CreateLoad(valueTy, argIt);
+
+  if (is64Ty)
+    value = m_builder->CreateBitCast(value, m_builder->getDoubleTy());
+
+  Type *gpuVaAsPtrTy = Type::getInt8PtrTy(m_builder->getContext(), SPIRAS_Uniform);
+  auto gpuVaAsPtr = m_builder->CreateIntToPtr(gpuVa, gpuVaAsPtrTy);
+  // Create GEP to get the byte address with byte offset
+  gpuVaAsPtr = m_builder->CreateGEP(m_builder->getInt8Ty(), gpuVaAsPtr, offset);
+  Type *gpuVaPtrTy = is64Ty ? Type::getDoublePtrTy(m_builder->getContext(), SPIRAS_Uniform)
+                            : Type::getFloatPtrTy(m_builder->getContext(), SPIRAS_Uniform);
```

Comment:
If you use `valueTy == double`, this can be simplified. Also, the address space should be `SPIRAS_Global` since this is not a fat pointer.

Reply:
The given argument is not an actual VA, but the result from `AmdExtGetBaseAddrFromResource`, the resource has its address space originally `SPIRAS_Uniform`. I may change the variable names and comment to reflect that.

I've done it this way because it seems that we have no way to retrieve the VA of a resource at this point, so I use a `GEP` to the resource and cast it to uint64, then later cast it back to further access. Any better suggestion is welcomed!

Reply:
Okay, I was wondering about the implementation of `AmdExtGetBaseAddrFromResource` as well. What's the input type of `AmdExtGetBaseAddrFromResource`? I can't seem to find it in gpurt. But conceptually, if the input is a pointer to `addrspace(7)`, then a legitimate way to get the corresponding address in LLVM IR is to first `addrspacecast` to `addrspace(1)` (global) and then `ptrtoint` to get the `u64`.

The atomic function should then do everything in `addrspace(1)`.

Reply:
I should mention that for this to work, you'll need to support `addrspacecast` in the `PatchBufferOp` pass, but that should be relatively simple and is likely going to be useful for other things.

Reply:
The gpurt part has not been published yet, I've sent you an email internally to illustrate that.